### PR TITLE
Proposal: Software Defined Networking

### PR DESCRIPTION
**Pull request by mfernd**

### CODE REVIEW

File: docs/proposals/network/README.md
Comment (sameo) [inline docs/proposals/network/README.md]: Pretty nice. Keep in mind that simplicity and is one of our goals here, performance is not important for now.

File: docs/proposals/network/README.md
Comment (sameo) [inline docs/proposals/network/README.md]: br ?

File: docs/proposals/network/README.md
Comment (sameo) [inline docs/proposals/network/README.md:41]: A third option is a binary to be called directly by the agent. This is the CNI approach.

File: docs/proposals/network/README.md
Comment (sameo) [inline docs/proposals/network/README.md]: kube-proxy is not really a feature or a requirement. Please remove that from the scope, or describe which kube-proxy features or characteristics are part of your scope.

File: docs/proposals/network/README.md
Comment (sameo) [inline docs/proposals/network/README.md]: FWIW, this should not be part of your final version. You have to make a decision about that as part of this architecture document.

File: docs/proposals/network/README.md
Comment (mfernd) [reply docs/proposals/network/README.md:41]: Previously, we didn't understood how the CNI approach worked, but now we think it's more comprehensible

You can see our approach in the last commits:

- for the [definition](https://github.com/dev-sys-do/orka/blob/8e83b9971ca8364c6bd3aeec86938dca99cfbc0b/docs/proposals/network/README.md?plain=1#L26)
- and the [plugin implementation](https://github.com/dev-sys-do/orka/blob/8e83b9971ca8364c6bd3aeec86938dca99cfbc0b/docs/proposals/network/README.md?plain=1#L74)

File: docs/proposals/network/README.md
Comment (mfernd) [reply docs/proposals/network/README.md]: Useless `br` (line break) sorry, it's removed

File: docs/proposals/network/README.md
Comment (mfernd) [reply docs/proposals/network/README.md]: Yes, of course, we need something that is working first

We keeped the examples section

File: docs/proposals/network/README.md
Comment (mfernd) [reply docs/proposals/network/README.md]: We have specified with a `Scope` and `Out of scope`, tell us if it's better

**Comment by mfernd** [comment]
We rewrited the proposal with the CNI specifications in mind.  
(As a set of plugins to execute to manage network interfaces, if we understand it correctly.)

We've also drafted a brief summary of the CNI specifications, for our use case, and how we should implement it.

A more precise configuration for our CNI plugin should also be written to accomplish the first goal of our component ([intra-pod networking](https://github.com/dev-sys-do/orka/blob/0de0b667c1ade7ea3882f8db528c22c7e001dcbb/docs/proposals/network/README.md?plain=1#L116)).

File: docs/proposals/network/README.md
Comment (sameo) [inline docs/proposals/network/README.md]:
```suggestion
### Glossary
```

File: docs/proposals/network/README.md
Comment (sameo) [inline docs/proposals/network/README.md]:
```suggestion
The *plugins* (= binaries) must be installed on every node in the cluster.
```

File: docs/proposals/network/README.md
Comment (sameo) [inline docs/proposals/network/README.md]: If you specify in your glossary section that a plugin is a binary, you don't need the (= binaries) addition.

File: docs/proposals/network/README.md
Comment (sameo) [inline docs/proposals/network/README.md]:
```suggestion
- *runtime* is the program responsible for executing CNI plugins.
- *plugin* is a program that applies a specified network configuration.
```

Also, what is the network configuration applied on? Is it node specific, container specific?

File: docs/proposals/network/README.md
Comment (sameo) [inline docs/proposals/network/README.md]: I don't see how that's related to `orka`

File: docs/proposals/network/README.md
Comment (sameo) [inline docs/proposals/network/README.md]:
```suggestion
* RBAC and Network Policies 
* Monitoring network traffic
* Load balancing (e.g. service)
* Load balancer (e.g. klipper, metallb)
```

If the scope of your component ever changes, then you can change that section.

File: docs/proposals/network/README.md
Comment (sameo) [inline docs/proposals/network/README.md:65]: So you're planning to implement a CNI-compliant plugin? That's fine with me, but this section should be very explicit about it. Something like:

```
The `orka` SDN implements the [`CNI`](https://www.cni.dev/docs/spec/) specification. 
It is a `CNI` plugin and must be called by e.g. the node Agent through the `CNI` specified
command line interface. It configure workloads networking according to a set of `CNI`
compliant configuration files, as described in the following sections.


```

File: docs/proposals/network/README.md
Comment (mfernd) [reply docs/proposals/network/README.md]: That's not relevant anymore, I will remove this section, okay

It was to see an example of a minimal k8s and compare it to our side, but yes it's not necessary

File: docs/proposals/network/README.md
Comment (mfernd) [reply docs/proposals/network/README.md:65]: We did say that in the [Definition section](https://github.com/dev-sys-do/orka/blob/57dc5fc4101f50b1bfdad128af6968fdcda90ec2/docs/proposals/network/README.md#definition), but maybe it was not clear enough about the link to orka SDN.

I've used your paragraph and modified it slightly to update this section ([Definition](https://github.com/dev-sys-do/orka/blob/57dc5fc4101f50b1bfdad128af6968fdcda90ec2/docs/proposals/network/README.md#definition)).

I don't know if we should explain the goal of the SDN in the API section (instead of the Definition section), I prefer the opposite to reserve the API section with more explicit explanation?

Comment (sameo) [reply docs/proposals/network/README.md:65]: Yes, this is fine to me. It's clearer now, thanks.

File: docs/proposals/network/README.md
Comment (sameo) [inline docs/proposals/network/README.md]: I guess this is the only thing left for you to clarify until we can merge that PR.

**Comment by mfernd** [comment]
For the [API section](https://github.com/dev-sys-do/orka/pull/4/files#diff-3725324f4aca8745847906a1d96a7853cdbe9497da33a45d9ed67830dd93408aR63), we defined a **provisioning** & **execution** phase to better differentiate between the two of them.

What is left to do (in my opinion):

- [x] Define the configuration file for our plugin
- [x] ~~Add examples for `CHECK` and `VERSION` methods~~ removed
- [x] ~~See what we could do for DNS routing/resolution(?)~~ removed

**Comment by sameo** [comment]
> For the [API section](https://github.com/dev-sys-do/orka/pull/4/files#diff-3725324f4aca8745847906a1d96a7853cdbe9497da33a45d9ed67830dd93408aR63), we defined a **provisioning** & **execution** phase to better differentiate between the two of them.
> 
> What is left to do (in my opinion):

I agree with the list, but we can merge this PR once you have defined the config file. The rest can be defined as a follow up.
So, please focus on the first bullet below and we'll go ahead and merge.


> * [ ]  Define the configuration file for our plugin
> * [ ]  Add examples for `CHECK` and `VERSION` methods
> * [ ]  See what we could do for DNS routing/resolution(?)

**Comment by mfernd** [comment]
Now we have a configuration file for our CNI plugin, I think it's ready so I marked the PR as "ready for review".

What do we need to do before merging it?

**Review by sameo — APPROVED** [review]

**Comment by sameo** [comment]
Please squash all 16 commits into one and I'll merge it.

**Comment by sameo** [comment]
I went ahead and squashed the commits. The original branch is here: https://github.com/sameo/orka/tree/proposals/network-orig

**Comment by mfernd** [comment]
okay sorry for the late response and thank you

**Pull request merged**
